oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,293 @@
+.. _Google Docs: https://docs.google.com
+
+
+Decisions
+---------
+
+.. image:: oep-0019/developer_docs.png
+   :alt: A diagram that shows the different types of documentation and their
+    relationships. This is further described below.
+
+Agile Documentation Needs
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With Agile development and `Clean Code`_, we can minimize documentation (and its maintainence) since code should
+be readable and self-documenting. However, for long-term maintainence of a platform, the following developer
+documentation are still needed:
+
+1. **System-wide Decisions and Best Practices (OEPs)** are necessary to establish coherency
+   and efficiency with minimal centralized governance across a long-term maintained platform.
+   
+   * As described in OEP-1_, system-wide technical decisions, best practices, and development
+     processes are captured in the `open-edx-proposals`_ repo.
+
+2. **Architecture Decision Records (ADRs)** are local lightweight documents of
+   technical decisions co-located with its corresponding code. They are needed to 
+   understand the rationale behind the code and the archeological path to its current state. 
+   These records prove to be an `invaluable technique`_ for engineering teams to onboard
+   new members, to understand discovered historical traces, to refer back to past decisions,
+   and to support external oversight.
+   
+   * Note: Since ADRs are inherently a historical trail of technical decisions, by design,
+     they do not need ongoing maintainence.
+
+   * Example: A `local decisions`_ folder will be created as a top-level folder in a repo
+     and/or as subfolders of a repo, depending on the size of the repo's codebase. The goal
+     is to keep the decisions in as close proximity to the code as possible for their discovery
+     and usage.
+
+3. **READMEs** provide a top-level summary and intentional purpose of the repo or app.
+
+   * The README.rst includes a description of the *responsibility* of that component. This is important
+     so that future development remains within the boundaries of the intention of the author rather than
+     the component evolving in incongruent ways. In the event that the responsibility of the component
+     changes, the README.rst needs to be updated.
+
+   * The README may also serve as an index to the published docs of the component.
+
+4. **How-To Documentation** provides step-by-step instruction in enabling and using a feature.
+
+   * These will be stored in Github co-located with the relevant code.
+
+5. **API Documentation** provide information on interfaces for integrating with or enhancing the platform.
+   These should be automatically generated, as possible. For example:
+
+   * REST API interfaces will be automatically generated using Swagger_.
+   * The Open edX Learning Exchange (OLX) format should be automatically generated.
+   * Plugin interfaces (e.g., Proctoring integration) should be automatically generated using Sphinx_.
```

Comment:
I think this section should also explicitly call out generated documentation from python docstrings. Generating that documentation, publishing it to read-the-docs, and making it linkable between repos using `intersphinx` is something that should be standard, in my opinion.

Reply:
I added [another bullet point](https://github.com/edx/open-edx-proposals/pull/74/files#diff-63e65fcef1f4cd3676f373d1bea15029R209) for Public (Python and JS) APIs.

Thanks for the pointer to `intersphinx`.  Once we invest further time and experiment with API docs, we can update this OEP or create a new OEP with further detail.